src/BatteryStateOfChargeService.cpp

```diff
@@ -43,21 +44,65 @@ BatteryStateOfChargeService::~BatteryStateOfChargeService()
 
 double BatteryStateOfChargeService::totalAmpHoursUsed() const
 {
-    return 0.0;
+    return BATTERY_AMP_HOUR_CAPACITY - totalAH_;
 }
 
 bool BatteryStateOfChargeService::isCharging() const
 {
-    return false;
+    return isCharging_;
 }
 
+
 QTime BatteryStateOfChargeService::timeWhenChargedOrDepleted() const
 {
-    return QTime::currentTime();
+    QTime ntime(0,0,0,0);
+    if (time.isNull())
+        return ntime;
+    else
+        return time;
 }
 
 void BatteryStateOfChargeService::addData(const BatteryData& batteryData)
 {
-    Q_UNUSED(batteryData);
-    // Update your variables here.
+    if (prTime_.isNull())
+    {
+        totalAH_=BATTERY_AMP_HOUR_CAPACITY*initialStateOfChargePercent_/100;
+        prTime_=batteryData.time;
+    }
+    deltaT_=prTime_.msecsTo(batteryData.time);
+    deltaC_=(batteryData.current+preC_)/2;
+    amph_=deltaT_*deltaC_/3600000;
+    totalAH_+=amph_;
+
+    if (batteryData.current < 0){
+        isCharging_=true;
+        deltaAH_=-1*(BATTERY_AMP_HOUR_CAPACITY-totalAH_);
+
+    }
+    else{
+        isCharging_=false;
+        deltaAH_=totalAH_;
+    }
+    deltaAH_ /= deltaC_;
+
+    int hours = deltaAH_;
+    int min = (deltaAH_-hours)*60;
+    int sec =(deltaAH_-hours-(min/60))*60;
+    int ms = (deltaAH_-hours-min/60-sec/3600)*1000;
+    time.setHMS(hours, min, sec, ms);
```

Comment:
This section could be done cleaner by using the `addMSecsTo` function, and adding milliseconds to a `QTime(0, 0)` object.